File: node/src/serde/contract-hash.ts
import { createHash } from "node:crypto";
import protobuf from "protobufjs";
import type { SchemaPair, Serializer } from "./serializer";

// HASH_PREFIX tags the wire descriptor algorithm. It travels inside the hash
// string so a mesh running mixed algorithm generations never matches a hash
// computed under different rules.
const HASH_PREFIX = "v2";

type Cardinality = "opt" | "rep" | "map" | "req";

interface FieldDescriptor {
	n: number;
	c: Cardinality;
	t: string;
	k?: string;
}

interface OneofDescriptor {
	f: number[];
}

interface MessageDescriptor {
	f: FieldDescriptor[];
	o?: OneofDescriptor[];
}

interface EnumDescriptor {
	e: number[];
}

type ReferencedType = MessageDescriptor | EnumDescriptor;

// wireDescriptors holds the canonical wire descriptor of every Serializer that
// serde builds, keyed by identity. It is a side table rather than a Serializer
// member because the descriptor describes the protobuf Type behind the
// serializer, not the encode/decode contract other domains depend on.
const wireDescriptors = new WeakMap<object, string>();

// attachWireDescriptor binds a serializer to the canonical descriptor of its
// message. Called by every serializer factory in this module's domain.
//
// @internal — см. ./README.md
export function attachWireDescriptor(
	serializer: Serializer,
	type: protobuf.Type,
): void {
	wireDescriptors.set(serializer, canonicalMessageDescriptor(type));
}

// computeContractHash returns the routing identity of an (input, output) pair.
//
//	contractHash = "v2:" + hex(sha256(canon(input) + ":" + canon(output)))
//
// canon() is the language-neutral wire descriptor built by
// canonicalMessageDescriptor: field numbers, cardinality and types only. Field
// names are excluded, so a rename — wire-compatible in protobuf — keeps traffic
// flowing, while a number, type or cardinality change routes to a new identity.
//
// This is the SINGLE source of truth for contract hashing across the SDK.
// Runtime stores the hash opaque and does NOT recompute it.
//
// @public — см. ./README.md
export function computeContractHash(pair: SchemaPair): string {
	const input = wireDescriptor(pair.input);
	const output = wireDescriptor(pair.output);
	const digest = createHash("sha256")
		.update(`${input}:${output}`)
		.digest("hex");
	return `${HASH_PREFIX}:${digest}`;
}

// wireDescriptor returns the canonical descriptor string a serializer was built
// with — the exact bytes that feed the contract hash.
//
// @internal — см. ./README.md
export function wireDescriptor(serializer: Serializer): string {
	const descriptor = wireDescriptors.get(serializer);
	if (descriptor === undefined) {
		throw new Error(
			"serde: serializer has no wire descriptor — build the pair through buildSchemaPair",
		);
	}
	return descriptor;
}

// canonicalMessageDescriptor renders a protobuf message as the canonical JSON
// string that feeds the contract hash. Shape:
//
//	{"f":[{"n":1,"c":"opt","t":"string"}],
//	 "o":[{"f":[4,5]}],
//	 "r":{".pkg.Item":{"f":[...]},".pkg.Kind":{"e":[0,1]}}}
//
// `o` is omitted when the message has no oneof group, `r` when it references no
// other type. Full type names carry a leading dot.
//
// @internal — см. ./README.md
export function canonicalMessageDescriptor(type: protobuf.Type): string {
	const referenced: Record<string, ReferencedType> = {};
	const shape = messageShape(type, referenced);
	const doc: Record<string, unknown> = { f: shape.f };
	if (shape.o) {
		doc.o = shape.o;
	}
	if (Object.keys(referenced).length > 0) {
		doc.r = referenced;
	}
	return canonicalize(doc);
}

function messageShape(
	type: protobuf.Type,
	referenced: Record<string, ReferencedType>,
): MessageDescriptor {
	const fields = type.fieldsArray
		.map((field) => fieldDescriptor(field, referenced))
		.sort((a, b) => a.n - b.n);

	const oneofs = type.oneofsArray
		// A proto3 `optional` field is a single-member synthetic oneof in the
		// descriptor model; on the wire it is a plain singular field.
		.filter((oneof) => !oneof.isProto3Optional)
		.map((oneof) => ({
			f: oneof.fieldsArray.map((field) => field.id).sort((a, b) => a - b),
		}))
		.filter((group) => group.f.length > 0)
		.sort((a, b) => (a.f[0] as number) - (b.f[0] as number));

	const shape: MessageDescriptor = { f: fields };
	if (oneofs.length > 0) {
		shape.o = oneofs;
	}
	return shape;
}

function fieldDescriptor(
	field: protobuf.Field,
	referenced: Record<string, ReferencedType>,
): FieldDescriptor {
	try {
		field.resolve();
	} catch (err) {
		throw new Error(
			`serde: resolve field ${field.fullName}: ${(err as Error).message}`,
		);
	}

	const t = typeRef(field, referenced);
	if (field instanceof protobuf.MapField) {
		return { n: field.id, c: "map", k: field.keyType, t };
	}
	const c: Cardinality = field.repeated
		? "rep"
		: field.required
			? "req"
			: "opt";
	return { n: field.id, c, t };
}

function typeRef(
	field: protobuf.Field,
	referenced: Record<string, ReferencedType>,
): string {
	const resolved = field.resolvedType;
	if (resolved === null) {
		return field.type;
	}

	const name = resolved.fullName;
	if (resolved instanceof protobuf.Enum) {
		if (!(name in referenced)) {
			referenced[name] = { e: enumNumbers(resolved) };
		}
		return `e:${name}`;
	}

	if (!(name in referenced)) {
		// Claim the slot before recursing: a cyclic reference resolves by name
		// against this placeholder instead of recursing forever.
		referenced[name] = { f: [] };
		referenced[name] = messageShape(resolved as protobuf.Type, referenced);
	}
	return `m:${name}`;
}

function enumNumbers(enm: protobuf.Enum): number[] {
	return [...new Set(Object.values(enm.values))].sort((a, b) => a - b);
}

// canonicalize produces the deterministic JSON string used for hashing.
// Rules:
//   - null / primitives → JSON.stringify (standard)
//   - arrays → preserve element order, recurse on each element
//   - objects → sort keys lexicographically, recurse on each value
//   - no whitespace, no trailing commas
//
// @internal — см. ./README.md
export function canonicalize(value: unknown): string {
	if (value === null || typeof value !== "object") {
		return JSON.stringify(value);
	}
	if (Array.isArray(value)) {
		return `[${value.map(canonicalize).join(",")}]`;
	}
	const obj = value as Record<string, unknown>;
	const keys = Object.keys(obj).sort();
	return `{${keys
		.map((k) => `${JSON.stringify(k)}:${canonicalize(obj[k])}`)
		.join(",")}}`;
}

File: go/internal/serde/contracthash.go
// Package serde derives the wire identity of a protobuf contract and moves
// application messages between their Go form and the two wire forms the runtime
// stores.
package serde

import (
	"bytes"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"sort"

	"google.golang.org/protobuf/reflect/protoreflect"
	"google.golang.org/protobuf/types/known/emptypb"
)

// HashPrefix tags the algorithm generation inside the hash string itself, so a
// mesh running mixed generations never matches a hash computed under different
// rules.
const HashPrefix = "v2"

// ContractHash is the routing identity of an (input, output) message pair:
//
//	"v2:" + hex(sha256(canon(input) + ":" + canon(output)))
//
// canon is the language-neutral wire descriptor built by Canonical: field
// numbers, cardinality and types only. Field names are excluded because
// renaming a field is wire-compatible in protobuf and must not reroute live
// traffic, while a number, type or cardinality change must.
func ContractHash(in, out protoreflect.MessageDescriptor) string {
	var buf bytes.Buffer
	buf.Write(Canonical(in))
	buf.WriteByte(':')
	buf.Write(Canonical(out))
	sum := sha256.Sum256(buf.Bytes())
	return HashPrefix + ":" + hex.EncodeToString(sum[:])
}

// EventContractHash is the identity of an event payload. An event has no reply,
// so google.protobuf.Empty stands in for the output half — the same shape a
// one-way method would have, rather than a second hashing rule to keep in sync.
func EventContractHash(md protoreflect.MessageDescriptor) string {
	return ContractHash(md, (&emptypb.Empty{}).ProtoReflect().Descriptor())
}

// Canonical renders one message as the canonical JSON descriptor that feeds the
// contract hash:
//
//	{"f":[{"c":"opt","n":1,"t":"string"}],
//	 "o":[{"f":[4,5]}],
//	 "r":{".pkg.Item":{"f":[...]},".pkg.Kind":{"e":[0,1]}}}
//
// "o" is absent when the message has no oneof group, "r" when it references no
// other type. Full type names carry a leading dot.
//
// @internal — см. ./README.md
func Canonical(md protoreflect.MessageDescriptor) []byte {
	refs := map[string]any{}
	doc := shape(md, refs)
	if len(refs) > 0 {
		doc["r"] = refs
	}
	return encode(doc)
}

// shape builds one message's descriptor, recording everything it reaches in
// refs. The map is shared across the whole traversal, so reachability is
// transitive and every type is described exactly once.
func shape(md protoreflect.MessageDescriptor, refs map[string]any) map[string]any {
	fds := md.Fields()
	fields := make([]any, 0, fds.Len())
	for i := range fds.Len() {
		fields = append(fields, field(fds.Get(i), refs))
	}
	sort.Slice(fields, func(a, b int) bool {
		return fields[a].(map[string]any)["n"].(int) < fields[b].(map[string]any)["n"].(int)
	})

	doc := map[string]any{"f": fields}
	if groups := oneofs(md); len(groups) > 0 {
		doc["o"] = groups
	}
	return doc
}

// oneofs collects the real oneof groups. A proto3 `optional` field is a
// single-member synthetic oneof in the descriptor model but a plain singular
// field on the wire, so it is skipped.
func oneofs(md protoreflect.MessageDescriptor) []any {
	ods := md.Oneofs()
	groups := make([]any, 0, ods.Len())
	for i := range ods.Len() {
		od := ods.Get(i)
		if od.IsSynthetic() {
			continue
		}
		fds := od.Fields()
		nums := make([]int, 0, fds.Len())
		for j := range fds.Len() {
			nums = append(nums, int(fds.Get(j).Number()))
		}
		if len(nums) == 0 {
			continue
		}
		sort.Ints(nums)
		groups = append(groups, map[string]any{"f": nums})
	}
	sort.Slice(groups, func(a, b int) bool {
		return groups[a].(map[string]any)["f"].([]int)[0] < groups[b].(map[string]any)["f"].([]int)[0]
	})
	return groups
}

func field(fd protoreflect.FieldDescriptor, refs map[string]any) map[string]any {
	if fd.IsMap() {
		return map[string]any{
			"n": int(fd.Number()),
			"c": "map",
			"k": typeRef(fd.MapKey(), refs),
			"t": typeRef(fd.MapValue(), refs),
		}
	}
	card := "opt"
	switch fd.Cardinality() {
	case protoreflect.Repeated:
		card = "rep"
	case protoreflect.Required:
		card = "req"
	}
	return map[string]any{
		"n": int(fd.Number()),
		"c": card,
		"t": typeRef(fd, refs),
	}
}

// typeRef names the field's type and records the referenced message or enum.
func typeRef(fd protoreflect.FieldDescriptor, refs map[string]any) string {
	switch fd.Kind() {
	case protoreflect.EnumKind:
		ed := fd.Enum()
		name := "." + string(ed.FullName())
		if _, seen := refs[name]; !seen {
			refs[name] = map[string]any{"e": enumNumbers(ed)}
		}
		return "e:" + name
	case protoreflect.MessageKind, protoreflect.GroupKind:
		nested := fd.Message()
		name := "." + string(nested.FullName())
		if _, seen := refs[name]; !seen {
			// Claim the slot before recursing so a cycle resolves by name
			// against this placeholder instead of recursing forever.
			refs[name] = map[string]any{"f": []any{}}
			refs[name] = shape(nested, refs)
		}
		return "m:" + name
	default:
		return fd.Kind().String()
	}
}

func enumNumbers(ed protoreflect.EnumDescriptor) []int {
	vals := ed.Values()
	seen := make(map[int]struct{}, vals.Len())
	nums := make([]int, 0, vals.Len())
	for i := range vals.Len() {
		n := int(vals.Get(i).Number())
		if _, dup := seen[n]; dup {
			continue
		}
		seen[n] = struct{}{}
		nums = append(nums, n)
	}
	sort.Ints(nums)
	return nums
}

// encode writes the deterministic JSON the hash is taken over: object keys
// sorted at every level (encoding/json sorts map keys), array order preserved,
// no whitespace and no HTML escaping — escaping would make `<` in a type name
// produce bytes no other SDK produces.
func encode(v any) []byte {
	var buf bytes.Buffer
	enc := json.NewEncoder(&buf)
	enc.SetEscapeHTML(false)
	if err := enc.Encode(v); err != nil {
		// The document is built here from maps of strings, ints and slices of
		// the same. A value outside that set is unreachable.
		panic(fmt.Sprintf("serde: encode canonical descriptor: %v", err))
	}
	return bytes.TrimRight(buf.Bytes(), "\n")
}
